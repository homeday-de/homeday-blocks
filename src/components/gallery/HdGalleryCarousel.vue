<template>
  <div
    :class="{
      'gallery-carousel': true,
      isUsingMouse,
      withPagerInside: pagerInside,
    }"
    @keydown="setUsingMouse(false)"
    @mousedown="setUsingMouse(true)"
  >
    <div
      class="gallery-carousel__wrap"
      tabindex="0"
      @keydown.self="maybeSelectItem"
      @blur.self="setUsingMouse(false)"
    >
      <flickity
        ref="flickity"
        :options="flickityOptions"
      >
        <div
          v-for="(item, i) in items"
          :key="i"
          :class="{
            'gallery-carousel__item': true,
            'isActive': shouldShowActiveState(i),
          } "
        >
          <div
            :style="sizerStyles"
            class="gallery-carousel__item__sizer"
          />

          <!-- the item.thumbnail field is used as default value for the item image -->
          <!-- IE11 uses this value only because do not support the picture element -->
          <picture class="gallery-carousel__item__picture">
              <source v-for="(source, media) in item.thumbnailPictureSources"
                :key="media"
                :media="`(${media})`" :srcset="source"
              >
            <img :src="item.thumbnail" :alt="item.caption" :srcset="item.thumbnailSrcSet">
          </picture>

        </div>
      </flickity>
      <div class="gallery-carousel__pager">
        <HdPager
          v-if="pagerItemsCount"
          :count="pagerItemsCount"
          :white="pagerInside"
          v-model="currentIndex"
        />
      </div>
    </div>
  </div>
</template>

<script>
// eslint-disable-next-line import/extensions
import Flickity from 'vue-flickity';
import onResize from 'homeday-blocks/src/services/on-resize';
import HdPager from 'homeday-blocks/src/components/HdPager.vue';

export default {
  name: 'HdGalleryCarousel',
  components: {
    Flickity,
    HdPager,
  },
  props: {
    value: {
      type: Number,
      default: 0,
    },
    items: {
      type: Array,
      default: () => [],
    },
    pagerInside: {
      type: Boolean,
      default: false,
    },
    aspectRatio: {
      type: Number,
      default: 16 / 9,
    },
    disableKeyEvents: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      flickityOptions: {
        prevNextButtons: false,
        pageDots: false,
        groupCells: true,
        contain: true,
        accessibility: false,
      },
      currentIndex: 0,
      pagerItemsCount: 0,
      valueChanged: false,
      isUsingMouse: false,
    };
  },
  computed: {
    sizerStyles() {
      return {
        paddingTop: `${100 / this.aspectRatio}%`,
      };
    },
  },
  watch: {
    // We need to handle a case where a user clicks on the pager
    currentIndex(currentIndex) {
      if (currentIndex === this.$refs.flickity.selectedIndex()) {
        return;
      }

      this.$refs.flickity.select(currentIndex);
    },
    value() {
      this.valueChanged = true;
      this.slideCellIntoView();
    },
  },
  mounted() {
    this.$nextTick(() => {
      this.updatePagerItemsCount();
      this.$refs.flickity.on('change', this.updateCurrentIndex);
      this.$refs.flickity.on('staticClick', this.onStaticClick);
    });
    this.addResizeEvents();
  },
  beforeDestroy() {
    this.removeResizeEvents();
    this.$refs.flickity.off('change', this.updateCurrentIndex);
    this.$refs.flickity.off('staticClick', this.onStaticClick);
  },
  methods: {
    updatePagerItemsCount() {
      this.pagerItemsCount = this.$refs.flickity.slides().length;
    },
    delayedUpdatePagerItemsCount() {
      this.$nextTick(() => {
        this.updatePagerItemsCount();
        this.updateCurrentIndex();
      });
    },
    updateCurrentIndex(itemIndex) {
      this.currentIndex = itemIndex;
      this.$emit('input', itemIndex);
    },
    onStaticClick(event, pointer, cellElement, cellIndex) {
      if (typeof cellIndex === 'undefined') {
        return;
      }

      this.$emit('itemClick', cellIndex);
      this.triggerSelectedItem(cellIndex);
    },
    triggerSelectedItem(cellIndex) {
      this.valueChanged = true;
      this.$emit('input', cellIndex);
    },
    shouldShowActiveState(index) {
      if (!this.valueChanged) {
        return false;
      }

      return index === this.value;
    },
    addResizeEvents() {
      onResize.onDebounced(this.delayedUpdatePagerItemsCount);
    },
    removeResizeEvents() {
      onResize.offDebounced(this.delayedUpdatePagerItemsCount);
    },
    maybeSelectItem(e) {
      if (this.disableKeyEvents) {
        return;
      }

      if (['ArrowDown', 'ArrowRight'].includes(e.key)) {
        const currentIndex = this.value;

        if (currentIndex === this.$refs.flickity.cells().length - 1) {
          this.triggerSelectedItem(0);
          return;
        }

        this.triggerSelectedItem(currentIndex + 1);
      } else if (['ArrowUp', 'ArrowLeft'].includes(e.key)) {
        const currentIndex = this.value;

        if (currentIndex === 0) {
          this.triggerSelectedItem(this.$refs.flickity.cells().length - 1);
          return;
        }

        this.triggerSelectedItem(currentIndex - 1);
      }
    },
    slideCellIntoView() {
      const slidesCount = this.$refs.flickity.slides().length;
      const cellsCount = this.$refs.flickity.cells().length;
      const cellsPerSlide = Math.round(cellsCount / slidesCount);
      const targetSlide = Math.floor(this.value / cellsPerSlide);

      this.$refs.flickity.select(targetSlide);
    },
    setUsingMouse(usingMouse) {
      this.isUsingMouse = usingMouse;
    },
  },
};
</script>

<style lang="scss">
@import 'homeday-blocks/src/styles/mixins.scss';

.gallery-carousel {
  $_root: &;
  position: relative;

  &.isUsingMouse * {
    outline: 0;
  }

  &__wrap {
    overflow: hidden;

    @media (min-width: $break-tablet) {
      padding: $stack-m 0;
    }

    .flickity-viewport {
      overflow: visible;
    }
  }

  &__pager {
    @media (min-width: $break-tablet) {
      margin-top: $stack-m;
    }

    #{$_root}.withPagerInside & {
      position: absolute;
      bottom: 0px;
      width: 100%;
      background: linear-gradient(to top, rgba($nevada, .45), rgba($nevada, 0));

      @media (min-width: $break-tablet) {
        display: none;
      }
    }
  }

  &__item {
    position: relative;
    width: 100%;
    margin-right: $inline-xs;
    cursor: pointer;
    border-radius: 2px;
    overflow: hidden;
    transition: box-shadow $time-s ease-in-out;
    background-color: $wild-sand;
    display: flex;
    align-items: center;
    justify-content: center;

    @media (min-width: $break-tablet) {
      width: calc(100% / 5);
      margin-right: $inline-m;
      box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.25);

      &:hover,
      &:focus,
      &.isActive {
        box-shadow: 0 5px 7px -3px rgba(0, 0, 0, 0.5);
        @media (min-width: $break-tablet) {
          box-shadow: 0 6px 11px -3px rgba(0, 0, 0, 0.45);
        }
      }
    }

    @media (min-width: $break-desktop) {
      width: calc(100% / 7);
      margin-right: $inline-m;
    }

    img {
      width: 100%;
      height: 100%;
      object-fit: cover;
    }

    &__picture {
      position: absolute;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      display: flex;
      align-items: center;
      justify-content: center;
    }
  }
}
</style>
