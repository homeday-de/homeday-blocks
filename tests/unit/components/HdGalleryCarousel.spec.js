import { mount } from '@vue/test-utils';
import HdGalleryCarousel from '@/components/gallery/HdGalleryCarousel.vue';
import ITEMS from '@/stories/mocks/GALLERY_ITEMS';

describe('HdGalleryCarousel', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(HdGalleryCarousel, {
      propsData: {
        items: ITEMS,
      },
    });
  });

  it('renders component', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
