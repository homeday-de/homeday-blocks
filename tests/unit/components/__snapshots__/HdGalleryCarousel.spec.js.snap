// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdGalleryCarousel renders component 1`] = `
<div class="gallery-carousel">
  <div tabindex="0" class="gallery-carousel__wrap">
    <div class="flickity-enabled">
      <div class="flickity-viewport" style="height: 0px;">
        <div class="flickity-slider" style="left: 0px; transform: translateX(NaN%);">
          <div class="gallery-carousel__item is-selected" style="position: absolute;">
            <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
            <picture class="gallery-carousel__item__picture">
              <source media="(max-width: 900px)" srcset="https://dummyimage.com/768x432?text=1+thumb+M+to+TP"> <img src="https://dummyimage.com/154x86?text=1+thumb+TL+to+D" alt="Picture Sources Test"></picture>
          </div>
          <div class="gallery-carousel__item is-selected" style="position: absolute;">
            <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
            <picture class="gallery-carousel__item__picture">
              <source media="(max-width: 900px)" srcset="https://dummyimage.com/768x432?text=2+thumb+M+to+TP"> <img src="https://dummyimage.com/154x86?text=2+thumb+TL+to+D" alt="Picture Sources Test"></picture>
          </div>
          <div class="gallery-carousel__item is-selected" style="position: absolute;">
            <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
            <picture class="gallery-carousel__item__picture"> <img src="https://dummyimage.com/154x86?text=3+thumb+default+density" alt="Image Source Set test" srcset="https://dummyimage.com/308x172?text=3+thumb+DPR2x 2x, https://dummyimage.com/462x258?text=3+thumb+DPR3x 3x"></picture>
          </div>
          <div class="gallery-carousel__item is-selected" style="position: absolute;">
            <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
            <picture class="gallery-carousel__item__picture">
              <source media="(max-width: 900px)" srcset="https://dummyimage.com/768x432?text=4+thumb+M+to+TP+DPR1x 1x, https://dummyimage.com/1536x864?text=4+thumb+TP+DPI2x 2x, https://dummyimage.com/2304x1296?text=4+thumb+TP3x 3x"> <img src="https://dummyimage.com/306x172?text=4+thumb+default+density" alt="Image and Picture Source Sets test" srcset="https://dummyimage.com/612x344?text=4+thumb+DPR2x 2x, https://dummyimage.com/918x516?text=4+thumb+DPR3x 3x"></picture>
          </div>
          <div class="gallery-carousel__item is-selected" style="position: absolute;">
            <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
            <picture class="gallery-carousel__item__picture"> <img src="https://dummyimage.com/306x172?text=6" alt="Cellar"></picture>
          </div>
          <div class="gallery-carousel__item is-selected" style="position: absolute;">
            <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
            <picture class="gallery-carousel__item__picture"> <img src="https://dummyimage.com/2306x3172?text=5+thumb+uneven+image" alt="Uneven image"></picture>
          </div>
          <div class="gallery-carousel__item is-selected" style="position: absolute;">
            <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
            <picture class="gallery-carousel__item__picture"> <img src="https://dummyimage.com/306x172?text=7" alt="__vue_devtool_undefined__"></picture>
          </div>
          <div class="gallery-carousel__item is-selected" style="position: absolute;">
            <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
            <picture class="gallery-carousel__item__picture"> <img src="https://dummyimage.com/306x172?text=8" alt="__vue_devtool_undefined__"></picture>
          </div>
          <div class="gallery-carousel__item is-selected" style="position: absolute;">
            <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
            <picture class="gallery-carousel__item__picture"> <img src="https://dummyimage.com/306x172?text=9" alt="Some title"></picture>
          </div>
          <div class="gallery-carousel__item is-selected" style="position: absolute;">
            <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
            <picture class="gallery-carousel__item__picture"> <img src="https://dummyimage.com/306x172?text=10" alt="Some title"></picture>
          </div>
          <div class="gallery-carousel__item is-selected" style="position: absolute;">
            <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
            <picture class="gallery-carousel__item__picture"> <img src="https://dummyimage.com/306x172?text=11" alt="Some title"></picture>
          </div>
          <div class="gallery-carousel__item is-selected" style="position: absolute;">
            <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
            <picture class="gallery-carousel__item__picture"> <img src="https://dummyimage.com/306x172?text=12" alt="Some title"></picture>
          </div>
        </div>
      </div>
    </div>
    <div class="gallery-carousel__pager">
      <!---->
    </div>
  </div>
</div>
`;
